scipy/optimize/least_squares.py

```diff
@@ -0,0 +1,657 @@
+                     "differencing will be used.")
+                sparsity = None
+            else:
+                sparsity = check_jac_sparsity(jac_sparsity, n, m)
+        else:
+            sparsity = None
+
+        def jac_wrapped(x, f):
+            J = approx_derivative(
+                fun, x, rel_step=diff_step, method=jac, f0=f, bounds=bounds,
+                args=args, kwargs=kwargs, sparsity=sparsity)
+
+            if issparse(J) and tr_solver == 'exact':
+                warn("Jacobian is converted to dense for tr_solver='exact' "
+                     "even though `jac_sparsity` is provided. Consider using "
+                     "'lsmr' trust-region solver or set `jac_sparsity` to "
+                     "None.")
+                J = J.toarray()
+            elif not issparse(J):
+                J = np.atleast_2d(J)
+
+            return J
+    else:
+        def jac_wrapped(x, _=None):
+            J = jac(x, *args, **kwargs)
+
+            if issparse(J):
+                if method == 'lm':
+                    warn("Jacobian is converted to dense for "
+                         "method='lm', consider using a different `method` "
+                         "or return dense Jacobian.")
+                    J = J.toarray()
+                elif tr_solver == 'exact':
+                    warn("Jacobian is converted to dense for "
+                         "tr_solver='exact', consider using 'lsmr' "
+                         "trust-region solver or return dense Jacobian.")
+                    J = J.toarray()
+                else:
+                    J = csr_matrix(J)
+            elif not isinstance(J, LinearOperator):
+                J = np.atleast_2d(J)
+
+            return J
+
+    J0 = jac_wrapped(x0, f0)
+    if len(J0.shape) != 2:
+        raise RuntimeError("`jac` must return at most 2-d array_like, "
+                           "sparse matrix or LinearOperator.")
+
+    if J0.shape != (m, n):
+        raise RuntimeError("The return value of `jac` has wrong shape.")
+
+    if isinstance(J0, LinearOperator):
+        if method == 'lm':
+            raise ValueError("method='lm' can't be used when `jac` "
+                             "returns LinearOperator.")
+
+        if tr_solver == 'exact':
+            raise ValueError("tr_solver='exact' can't be used when `jac` "
+                             "returns LinearOperator.")
+
+        if scaling == 'jac':
+            raise ValueError("scaling='jac' can't be used when `jac` "
+                             "return LinearOperator.")
+
+    if tr_solver is None:
+        if isinstance(J0, np.ndarray):
+            tr_solver = 'exact'
+        else:
+            tr_solver = 'lsmr'
+
+    if method == 'lm':
+        if jac == '2-point':
+            jac_wrapped = None
+        elif jac == '3-point':
+            jac_wrapped = None
+            warn("jac='3-point' works equivalently to '2-point' "
+                 "for 'lm' method.")
```

Comment:
Well, maybe indeed it's somewhat confusing. Perhaps I should separate validation of analytical and numerical Jacobians.